### PR TITLE
Keypoint updates

### LINKED ISSUE #1581: [FR] Add support for per-point confidence/visibility
### Proposal Summary

Add support for per-point confidence/visibility - currently Keypoint label class (which represents a set of points associated with an instance) is a flat list of (x, y) which doesn't provide additional per-point information. In addition, Coco import skips over the point which visibility = 0 (not visible/annotated) which can break the implicit mapping between joint -> point since num_joints != num_points and there's no information on which joint was skipped.

### Motivation

-   What is the use case for this feature?
This is needed to correctly import COCO keypoint dataset as well as filter sample based on per-point confidence...etc.

-   Why is this use case valuable to support for FiftyOne users in general?
Anyone using keypoint detection would probably need this since most of models outputs per-point confidence and there needs to be a way to distinguish in ground truth whether the point is visible or not + mapping between point to joint.

-   Why is this use case valuable to support for your project(s) or
    organization?
My project needs keypoint dataset - was hoping to leverage existing COCO dataset however the import breaks the mapping so any partially annotated keypoints (e.g. missing things) are useless.

-   Why is it currently difficult to achieve this use case? (please be as
    specific as possible about why related FiftyOne features and components are
    insufficient)
One way to workaround would be to only use fully annotated samples (no v=0), which would probably throw out majority of keypoint annotations.

### What areas of FiftyOne does this feature affect?

-   [x] App: FiftyOne application
-   [x] Core: Core `fiftyone` Python library
-   [ ] Server: FiftyOne server

### Details

(Use this section to include any additional information about the feature. If
you have a proposal for how to implement this feature, please include it here.)

We probably need to fix COCOObject::to_keypoints to not skip over v == 0 case and encode this visibility information to Keypoint by

Option 1. Expand KeypointsField to include confidence/visibility (not sure how it would support "optional-ness")
Option 2. Add a separate field "confidences"/"visibilities" in Keypoint class which maps to "points"

### Willingness to contribute

The FiftyOne Community encourages new feature contributions. Would you or
another member of your organization be willing to contribute an implementation
of this feature?

-   [ ] Yes. I can contribute this feature independently.
-   [x] Yes. I would be willing to contribute this feature with guidance from
        the FiftyOne community.
-   [ ] No. I cannot contribute this feature at this time.

I would love to contribute, however I'm not a Python expert + new to FiftyOne so would probably need some guidance.

### PR DESCRIPTION
**Pull request by brimoor**
Resolves #1581

Requires https://github.com/voxel51/eta/pull/556

## Change log

- [x] Implement keypoint skeletons as per #1563
- [x] Implement skeleton rendering in `draw_labels()`
- [x] Add a `filter_keypoints()` view stage that applies per-`point` filters to keypoint objects
- [x] Remove `visible` usage in favor of `NaN`-based filtering
- [x] Update `draw_labels()` code to gracefully handle NaN-valued keypoints
- [x] Documentation
- [x] Add keypoint skeleton (`label` and `edges`) rendering in the App
    - [x] Add an option to the App's settings menu to toggle whether skeletons are rendered
    - [x] Add skeleton labels and any `ListField` field values (whose length is `len(points)`) to the tooltip of individual points
- [x] Add support for custom per-point attributes on keypoints (occluded, etc)
- [x] Update App keypoint filtering logic:
    - [x] Add a `skeleton.label` filter that allows for filtering keypoints by their skeleton label
    - [x] Add a `confidence` filter that allows for filtering keypoints by `Keypoint.confidence`
- [x] ~~Convert `KeypointSkeleton` to just a dict?~~
- [x] ~~Add support for per-`label` skeletons in a field?~~

App implementation details
- The `Keypoint.points` filter logic should insert `NaN` values rather than removing the elements from the list (since the skeleton is defined based on list index position)
- Any `Keypoint.points` entry that contains a NaN coordinates should not be rendered in the App
- Any skeleton edges between one or more points with NaN coordinates should not be rendered
- The new `filter_keypoints()` view stage supports filtering by per-point confidences and per-point skeleton labels

## Example usage

```py
import numpy as np

import fiftyone as fo
import fiftyone.zoo as foz
from fiftyone import ViewField as F

dataset = foz.load_zoo_dataset("coco-2017", split="validation").clone()

model = foz.load_zoo_model("keypoint-rcnn-resnet50-fpn-coco-torch")
dataset.default_skeleton = model.skeleton

view = (
    dataset
    .match_labels(fields="ground_truth", filter=F("label") == "person")
    .limit(2)
)

view.apply_model(model, label_field="predictions")

# Add some custom attributes
for sample in view:
    for keypoint in sample["predictions_keypoints"].keypoints:
        keypoint["occluded"] = [np.random.randn() > 0 for _ in range(len(keypoint.points))]
        keypoint["age"] = np.random.randint(51)

    sample.save()

print(view.bounds("predictions_keypoints.keypoints.confidence"))
print(dataset.default_skeleton)

session = fo.launch_app(view)

session.view = view.filter_keypoints(
    "predictions_keypoints",
    filter=F("confidence") > 0.99999,
)

session.view = view.filter_keypoints(
    "predictions_keypoints",
    labels=["left eye", "right eye"],
)

view.draw_labels(
    "/tmp/keypoints",
    label_fields="predictions_keypoints",
    overwrite=True,
)
```

### Keypoint skeleton construction

```py
skeleton = fo.KeypointSkeleton(
    labels=[
        'nose',
        'left eye',
        'right eye',
        'left ear',
        'right ear',
        'left shoulder',
        'right shoulder',
        'left elbow',
        'right elbow',
        'left wrist',
        'right wrist',
        'left hip',
        'right hip',
        'left knee',
        'right knee',
        'left ankle',
        'right ankle',
    ],
    edges=[
        [11, 5, 3, 1, 0, 2, 4, 6, 12],
        [9, 7, 5, 6, 8, 10],
        [15, 13, 11, 12, 14, 16],
    ]
)
```

### CODE REVIEW

**Comment by brimoor** [comment]
@ehofesmann @benjaminpkane can you take a look at this PR and share any thoughts about the current keypoint interface?

I'm leaning towards *not* implementing the two outstanding backend TODO items in the PR description, but want to get your thoughts.

**Comment by brimoor** [comment]
@ehofesmann thanks for your detailed thoughts here.

Per your comments, I went ahead and nixed the per-label skeleton feature. I also didn't want to add it and was hoping for some confirmation bias :)

Regarding `Keypoint` and `Keypoints`, I agree that `Keypoint` being singular is confusing on its face, since, after all, it represents a collection of one or more points. However, since we do need a way to represent "points representing one person" versus "groups of points representing multiple people", our hand is basically forced. I'm okay with it though because this naming is semantically consistent with `Detection` vs `Detections`, for example.

Allowing `Keypoint.points` to be either a dict or a list is not something that we can do in FiftyOne right now. We could theoretically allow the user to choose between `Keypoint.points_list` or `keypoint.points_dict`, but I don't want to introduce that complexity right now. That would be similar to allowing users to choose between `Classification.label_str` and `Classification.target_int`, for example. Also thought about it, but not acting on it just yet.

We could add a factory method like `KeypointSkeleton.from_connections(labels, connections)` that would allow you construct a skeleton from a more human-readable data format like `connections=[("left eye", "right eye"), ...]`. But on the implementation side, this is always going to need to be converted to integer indices for the actual rendering. The choice basically comes down to whether the data model should be optimized for human-readability or computational-readiness/efficiency.

**Comment by brimoor** [comment]
Here's a direction we may want to go:

```py
class Keypoint(Label):
    label = fof.StringField()
    points = fof.ListField(fof.EmbeddedDocumentField(Point))

class Point(Label):
    x = fof.FloatField()
    y = fof.FloatField()
    confidence = fof.FloatField()
```

File: fiftyone/utils/labels.py [hunk 1]
```diff
@@ -341,3 +343,115 @@ def classification_to_detections(sample_collection, in_field, out_field):
             image[out_field] = fol.Detections(detections=[detection])
 
         sample.save()
+
+
+def filter_keypoints(sample_collection, field, expr=None, labels=None):
```
Comment (benjaminpkane) [inline fiftyone/utils/labels.py:348]: Shouldn't this proivde an `only_matches` option? If the confidence filter does not filter empty keypoints than it is better to do this all client side in the App.

Comment (brimoor) [reply fiftyone/utils/labels.py:348]: Hmm yeah that does make sense. If a filter omits all points in a `Keypoint`, remove the entire object.

I think I can add that

**Comment by benjaminpkane** [comment]
This is ready excluding some testing and confidence filtering. Once `only_matches` is added keypoint filtering I can finish.

**Comment by benjaminpkane** [comment]
keypoints

File: fiftyone/utils/labels.py (hunk 1 above)
Comment (brimoor) [reply fiftyone/utils/labels.py]: @benjaminpkane ok added the `only_matches` flag.

I'm thinking about converting this utility to a proper `FilterKeypoints(ViewStage)` and `SampleCollection.filter_keypoints()` view stage. Any objections?

Comment (benjaminpkane) [reply fiftyone/utils/labels.py]: I might prefer `FilterLabels` to support keypoints automatically. Perhaps in addition to `FilterKeypoints`. `FilterKeypoints` is somewhat specific but I suppose thats fine for now.

Comment (brimoor) [reply fiftyone/utils/labels.py]: I thought about merging this functionality into `FilterLabels`, but the problem is that will require introducing new optional kwargs, say `points_filter` and `points_labels`, that are only applicable when the field being filtered happens to be of type `Keypoint(s)`.

Comment (benjaminpkane) [reply fiftyone/utils/labels.py]: Ok. Cause to refactor `Keypoint` in the future

**Comment by benjaminpkane** [comment]
Ok, ready

**Comment by brimoor** [comment]
> Ok, ready

@benjaminpkane Almost there! I only saw two issues:
- In the grid view, when I use the `skeleton.label` filter widget, I get a `FilterLabels` stage (which isn't valid) but it should be a `FilterKeypoints` stage
- In the expanded modal, the tooltip does not appear when hovering over certain keypoints when `NaN` values are involved. Try loading the `filter_keypoints()` views in the code snippet in the PR description to see this

**Comment by benjaminpkane** [comment]
Fixed. 

> In the grid view, when I use the skeleton.label filter widget, I get a FilterLabels stage (which isn't valid) but it should be a FilterKeypoints stage

I was unable to reproduce the above. Let me know if it is still an issue

**Comment by brimoor** [comment]
> Fixed.
> 
> > In the grid view, when I use the skeleton.label filter widget, I get a FilterLabels stage (which isn't valid) but it should be a FilterKeypoints stage
> 
> I was unable to reproduce the above. Let me know if it is still an issue

LGTM, thanks! (probably I forgot to rebuild the App or something)

**Pull request merged**